java/algorithm/src/main/java/org/apache/arrow/algorithm/rank/VectorRank.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.arrow.algorithm.rank;
+
+import java.util.stream.IntStream;
+
+import org.apache.arrow.algorithm.sort.VectorValueComparator;
+import org.apache.arrow.memory.BufferAllocator;
+import org.apache.arrow.util.Preconditions;
+import org.apache.arrow.vector.IntVector;
+import org.apache.arrow.vector.ValueVector;
+
+/**
+ * Utility for calculating ranks of vector elements.
+ * @param <V> the vector type
+ */
+public class VectorRank<V extends ValueVector> {
+
+  private VectorValueComparator<V> comparator;
+
+  /**
+   * Vector indices.
+   */
+  private IntVector indices;
+
+  private final BufferAllocator allocator;
+
+  /**
+   * Constructs a vector rank utility.
+   * @param allocator the allocator to use.
+   */
+  public VectorRank(BufferAllocator allocator) {
+    this.allocator = allocator;
+  }
+
+  /**
```

Comment:
https://lemire.me/blog/2017/06/21/top-speed-for-top-k-queries/ seems to indicate that using a heap might be faster and probably more member efficient for extrema values.  Up to you if you want to change the algorithm.

Reply:
@emkornfield thanks for the suggested reference. 
A heap is good for the top k problem. However, this is a different problem from top k. 
We do not need to find top k elements. We only need to find the kth smallest value.

Using a top k algorithm for this problem is an over-kill: top k algorithm requires k * log(n) time, while this algorithm requires O(n).

Reply:
To some extent, this algorithm also finds the top k elements. However, it does not contain information about the relative order of the k elements.

Reply:
Like I said I think it is OK to keep this as is.  Note that the complexity analysis is the same for top-N is essentially the same.

Partition based is O(N). Using a heap is O(lg(k)*N + k).  For small k, lg(K) is effectively a constant.  Based on the benchmarks, it seems that do to processor effects it can be more efficient.  For thinks like median quick-select is probably better.

Reply:
Thanks for the clarification. I think the top n algorithm in the document is also interesting and practical.